gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java

```diff
@@ -224,12 +200,13 @@ public static ClientContext create(StubSettings settings) throws IOException {
     if (transportChannelProvider.needsCredentials() && credentials != null) {
       transportChannelProvider = transportChannelProvider.withCredentials(credentials);
     }
-    String endpoint =
-        getEndpoint(
-            settings.getEndpoint(),
-            settings.getMtlsEndpoint(),
-            settings.getSwitchToMtlsEndpointAllowed(),
-            new MtlsProvider());
+    EndpointContext endpointContext =
+        EndpointContext.newBuilder()
+            .setClientSettingsEndpoint(settings.getEndpoint())
+            .setMtlsEndpoint(settings.getMtlsEndpoint())
+            .setSwitchToMtlsEndpointAllowed(settings.getSwitchToMtlsEndpointAllowed())
+            .build();
+    String endpoint = endpointContext.resolveEndpoint();
```

Comment:
It maybe outside the scope of this PR, do we plan to pass the whole `EndpointContext` to `ApiCallContext` so that we can do the validation later?

Reply:
Yes, it'll be in a future PR.